ff/src/fields/models/fp/mod.rs

```diff
@@ -0,0 +1,1015 @@
+            // Location of the flag is the last byte of the serialized
+            // form of the field element.
+            let flag_location = output_byte_size - 1;
+
+            // At which byte is the flag located in the last limb?
+            let flag_location_in_last_limb = flag_location - (8 * (N - 1));
+
+            // Take all but the last 9 bytes.
+            let last_bytes = result_bytes.last_n_plus_1_bytes_mut();
+
+            // The mask only has the last `F::BIT_SIZE` bits set
+            let flags_mask = u8::MAX.checked_shl(8 - (F::BIT_SIZE as u32)).unwrap_or(0);
+
+            // Mask away the remaining bytes, and try to reconstruct the
+            // flag
+            let mut flags: u8 = 0;
+            for (i, (b, m)) in last_bytes.zip(&last_bytes_mask).enumerate() {
+                if i == flag_location_in_last_limb {
+                    flags = *b & flags_mask
+                }
+                *b &= m;
+            }
+            Self::deserialize(&result_bytes.as_slice()[..(N * 8)])
+                .ok()
+                .and_then(|f| F::from_u8(flags).map(|flag| (f, flag)))
+        }
+    }
+
+    #[inline]
+    fn square(&self) -> Self {
+        let mut temp = self.clone();
+        temp.square_in_place();
+        temp
+    }
+
+    fn square_in_place(&mut self) -> &mut Self {
+        P::square_in_place(self);
+        self
+    }
+
+    #[inline]
+    fn inverse(&self) -> Option<Self> {
+        P::inverse(&self)
+    }
+
+    fn inverse_in_place(&mut self) -> Option<&mut Self> {
+        if let Some(inverse) = self.inverse() {
+            *self = inverse;
+            Some(self)
+        } else {
+            None
+        }
+    }
+
+    /// The Frobenius map has no effect in a prime field.
+    #[inline]
+    fn frobenius_map(&mut self, _: usize) {}
+}
+
+impl<P: FpConfig<N>, const N: usize> PrimeField for Fp<P, N> {
+    type BigInt = BigInt<N>;
+    const MODULUS: Self::BigInt = P::MODULUS;
+    const GENERATOR: Self = P::GENERATOR;
+    const MODULUS_MINUS_ONE_DIV_TWO: Self::BigInt = P::MODULUS.divide_by_2_round_down();
+    const MODULUS_BIT_SIZE: u32 = P::MODULUS.const_num_bits();
+    const TRACE: Self::BigInt = P::MODULUS.two_adic_coefficient();
+    const TRACE_MINUS_ONE_DIV_TWO: Self::BigInt = Self::TRACE.divide_by_2_round_down();
+
+    #[inline]
+    fn from_bigint(r: BigInt<N>) -> Option<Self> {
+        P::from_bigint(r)
+    }
+
+    fn into_bigint(&self) -> BigInt<N> {
+        P::into_bigint(*self)
+    }
+}
+
+impl<P: FpConfig<N>, const N: usize> FftField for Fp<P, N> {
+    type FftConfig = P;
+
+    #[inline]
+    fn two_adic_root_of_unity() -> Self {
+        P::TWO_ADIC_ROOT_OF_UNITY
+    }
+
+    #[inline]
+    fn large_subgroup_root_of_unity() -> Option<Self> {
+        P::LARGE_SUBGROUP_ROOT_OF_UNITY
+    }
+
+    #[inline]
+    fn multiplicative_generator() -> Self {
+        P::GENERATOR
+    }
```

Comment:
Should these also be moved to constants?

Reply:
I think it is a good idea to move them into constants.

Reply:
Done!